### PR TITLE
CB-18507 - Prepared Images for upgrade.

### PR DESCRIPTION
**Pull request by vivprabu**
Adding boolean in upgradeV4Response to track what images are prepared for upgrade.

Added image id to message saved to StackStatus when CLUSTER_UPGRADE_PREPARATION_FINISHED status is saved.
Added boolean field to ImageInfoV4Response to indicate if an image id is already prepared.

Added Unit Tests to validate.

### CODE REVIEW

**Review by lacikaaa — CHANGES_REQUESTED** [review]
as we agreed on it, please store the prepared image related info in clustercomponent. don't use statusreason for anything like this

**Comment by vivprabu** [comment]
> as we agreed on it, please store the prepared image related info in clustercomponent. don't use statusreason for anything like this

done! can you please review again? @lacikaaa

**Review by lacikaaa — CHANGES_REQUESTED** [review]
I don't see any cleanup regarding prepared images. After the upgrade is done and the instances are replaced I think the prepared images should be cleaned

**Comment by vivprabu** [comment]
> I don't see any cleanup regarding prepared images. After the upgrade is done and the instances are replaced I think the prepared images should be cleaned

@lacikaaa addressed your comments

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/preparation/event/ClusterUpgradePreparationEvent.java [hunk 1]
```diff
@@ -3,27 +3,37 @@
...
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.sequenceiq.cloudbreak.cloud.model.ClouderaManagerProduct;
 import com.sequenceiq.cloudbreak.reactor.api.event.StackEvent;
 
+@JsonIgnoreProperties(ignoreUnknown = true)
 public class ClusterUpgradePreparationEvent extends StackEvent {
 
     private final Set<ClouderaManagerProduct> clouderaManagerProducts;
 
+    private final String imageId;
```
[the remark this answers is not in the document]
Comment (vivprabu) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/preparation/event/ClusterUpgradePreparationEvent.java:16]: done

File: core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterCommonService.java [hunk 2]
```diff
@@ -360,4 +367,31 @@ private void validateOperationOnStack(StackView stack, String operationDescripti
                     operationDescription));
         }
     }
+
+    public void updateOrSavePreparedClusterComponent(long stackId, String componentName,
```
[the remark this answers is not in the document]
Comment (vivprabu) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterCommonService.java]: done

File: core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterCommonService.java [hunk 3]
```diff
@@ -360,4 +367,31 @@ private void validateOperationOnStack(StackView stack, String operationDescripti
...
+            String imageId) throws IOException {
```
[the remark this answers is not in the document]
Comment (vivprabu) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterCommonService.java]: done

File: core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterCommonService.java [hunk 4]
```diff
@@ -360,4 +367,31 @@ private void validateOperationOnStack(StackView stack, String operationDescripti
...
+                ComponentType.CLUSTER_UPGRADE_PREPARED_IMAGES, componentName);
+        if (ObjectUtils.isEmpty(component)) {
+            createAndSaveClusterComponent(stackId, componentName, imageId);
+        } else {
+            updateClusterComponent(component, imageId);
+        }
+    }
+
+    private void createAndSaveClusterComponent(long stackId, String componentName, String imageId) {
+        Cluster clusterReference = clusterService.getClusterReference(stackId);
```
[the remark this answers is not in the document]
Comment (vivprabu) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterCommonService.java]: done

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/StackUpgradeOperations.java [hunk 5]
```diff
@@ -155,4 +171,18 @@ private boolean determineReplaceVmsParameter(Stack stack, Boolean replaceVms) {
             return upgradePreconditionService.notUsingEphemeralVolume(stack) && clusterDBValidationService.isGatewayRepairEnabled(stack.getCluster());
         }
     }
+
+    private List<String> getPreparedImagesList(long stackId) {
+        ClusterComponent clusterComponent = clusterComponentConfigProvider.getComponent(stackId, ComponentType.CLUSTER_UPGRADE_PREPARED_IMAGES,
```
[the remark this answers is not in the document]
Comment (vivprabu) [reply core/src/main/java/com/sequenceiq/distrox/v1/distrox/StackUpgradeOperations.java]: done

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/preparation/ClusterUpgradePreparationActions.java [hunk 6]
```diff
@@ -131,6 +136,12 @@ protected void doExecute(StackContext context, ClusterUpgradePreparationEvent pa
                 LOGGER.debug("Cluster upgrade preparation finish state started.");
                 Long resourceId = payload.getResourceId();
                 ResourceEvent resourceEvent = ResourceEvent.CLUSTER_UPGRADE_PREPARATION_FINISHED;
+                try {
+                    clusterCommonService.updateOrSavePreparedClusterComponent(resourceId,
+                            ResourceEvent.CLUSTER_UPGRADE_PREPARATION_FINISHED.name(), payload.getImageId());
```
[the remark this answers is not in the document]
Comment (vivprabu) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/preparation/ClusterUpgradePreparationActions.java]: done

**Comment by keyki** [comment]
@vivprabu could you please rebase your branch to top of the master so the swagger check will also pass ?

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeActions.java [hunk 7]
```diff
@@ -177,6 +182,8 @@ protected void doExecute(ClusterUpgradeContext context, ClusterUpgradeSuccess pa
                 StatedImage targetImage = getTargetImage(variables);
                 clusterUpgradeService.clusterUpgradeFinished(context.getStackId(), currentImage, targetImage);
                 stackImageService.removeImageByComponentName(context.getStackId(), TARGET_IMAGE);
+                clusterComponentUpdateService.deletePreparedClusterImages(context.getStackId(),
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeActions.java:185]: you are passing stack id again to a method which expects cluster id

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/StackUpgradeOperations.java [hunk 8]
```diff
@@ -162,4 +181,21 @@ private boolean determineReplaceVmsParameter(Stack stack, Boolean replaceVms) {
             return upgradePreconditionService.notUsingEphemeralVolume(stack) && clusterDBValidationService.isGatewayRepairEnabled(stack.getCluster());
         }
     }
+
+    private List<String> getPreparedImagesList(long stackId) {
+        List<String> preparedImages = null;
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/distrox/v1/distrox/StackUpgradeOperations.java:186]: this is a bit antipattern to return with null collection. please return with an empty one instead

File: core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterComponentUpdateService.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClusterComponentUpdateService.class);
+
+    @Inject
+    private ClusterService clusterService;
+
+    @Inject
+    private ClusterComponentConfigProvider clusterComponentConfigProvider;
+
+    public void updateOrSavePreparedClusterComponent(long stackId, ComponentType componentType,
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterComponentUpdateService.java:34]: I think `ComponentType componentType` parameter should be removed, as this method works only with this type of component only

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/StackUpgradeOperations.java (hunk 8 above)
Comment (vivprabu) [reply core/src/main/java/com/sequenceiq/distrox/v1/distrox/StackUpgradeOperations.java:186]: done

File: core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterComponentUpdateService.java (hunk 9 above)
Comment (vivprabu) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterComponentUpdateService.java]: done

**Comment by vivprabu** [comment]
> @vivprabu could you please rebase your branch to top of the master so the swagger check will also pass ?

rebased

File: core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterComponentUpdateService.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    public void deletePreparedClusterImages(long stackId, ComponentType componentType) {
+        LOGGER.debug("Deleting prepared images from Cluster Component for stack id: {}", stackId);
+        Optional<Long> clusterId = clusterService.findClusterIdByStackId(stackId);
+        if (clusterId.isPresent()) {
+            clusterComponentConfigProvider.deleteClusterComponentByClusterIdAndComponentType(clusterId.get(), componentType);
+            LOGGER.debug("Deleted prepared images from Cluster Component for stack id: {}", stackId);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (vivprabu) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterComponentUpdateService.java]: done

**Review by lacikaaa — APPROVED** [review]

**Review by lturcsanyi — APPROVED** [review]

**Pull request merged**